src/storage/compressed_tile.cpp

```diff
@@ -0,0 +1,301 @@
+    storage_manager.Release(backend_type, data);
+    data = NULL;
+
+    schema = catalog::Schema(columns);
+    tuple_length = schema.GetLength();
+    tile_size = num_tuple_slots * tuple_length;
+
+    LOG_TRACE("Size of the tile after compression: %zu", tile_size);
+
+    data = reinterpret_cast<char *>(
+        storage_manager.Allocate(backend_type, tile_size));
+
+    PL_ASSERT(data != NULL);
+    // zero out the data
+    PL_MEMSET(data, 0, tile_size);
+
+    for (oid_t i = 0; i < column_count; i++) {
+      bool is_inlined = schema.IsInlined(i);
+      size_t new_column_offset = schema.GetOffset(i);
+
+      if (GetCompressedType(i) == type::Type::INVALID) {
+        size_t old_column_offset = tile_schema->GetOffset(i);
+        auto old_column_type = tile_schema->GetColumn(i).GetType();
+
+        for (oid_t j = 0; j < num_tuple_slots; j++) {
+          type::Value old_value = tile->GetValueFast(
+              j, old_column_offset, old_column_type, is_inlined);
+          Tile::SetValueFast(old_value, j, new_column_offset, old_column_type,
+                             is_inlined);
+        }
+
+      } else {
+        auto new_column_type = schema.GetColumn(i).GetType();
+
+        for (oid_t j = 0; j < num_tuple_slots; j++) {
+          Tile::SetValueFast(new_columns[i][j], j, new_column_offset,
+                             new_column_type, is_inlined);
+        }
+      }
+    }
+
+    is_compressed = true;
+  }
+}
+
+std::vector<type::Value> CompressedTile::CompressColumn(
+    Tile *tile, oid_t column_id,
+    type::Type::TypeId compression_type = type::Type::TINYINT) {
+  oid_t num_tuples = tile->GetAllocatedTupleCount();
+  auto tile_schema = tile->GetSchema();
+  bool is_inlined = tile_schema->IsInlined(column_id);
+  size_t column_offset = tile_schema->GetOffset(column_id);
+  auto column_info = tile_schema->GetColumn(column_id);
+  auto column_type = column_info.GetType();
+  std::vector<type::Value> column_values(num_tuples);
+
+  for (oid_t i = 0; i < num_tuples; i++) {
+    column_values[i] =
+        tile->GetValueFast(i, column_offset, column_type, is_inlined);
+  }
+
+  std::vector<type::Value> actual_values(column_values);
+
+  std::sort(column_values.begin(), column_values.end(), CompareLessThanBool);
+
+  type::Value median = column_values[num_tuples / 2];
+  std::vector<type::Value> modified_values(num_tuples);
+
+  while (true) {
+    try {
+      type::Value min_diff =
+          column_values[0].Subtract(median).CastAs(compression_type);
+      type::Value max_diff =
+          column_values[num_tuples - 1].Subtract(median).CastAs(
+              compression_type);
+      for (oid_t k = 0; k < num_tuples; k++) {
+        modified_values[k] =
+            actual_values[k].Subtract(median).CastAs(compression_type);
+      }
+
+      LOG_TRACE("Base value selected = %s", median.ToString().c_str());
+      break;
+    } catch (Exception &e) {
+      if (column_type == (compression_type + 1)) {
```

Comment:
compression_type + 1 is the next bigger type. We only re-try compression when the next bigger type will not be equal to the current type. (This is because we compress only if the compression will lead to the use of a smaller data type.)